src/main/java/hw21/ProductsHandler.java

```diff
@@ -0,0 +1,57 @@
+package hw21;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class ProductsHandler {
+
+    public static List<Product> findExpensiveBooks(List<Product> products) {
+        Objects.requireNonNull(products);
+        return products.stream()
+                .filter(product -> product.isBook() && product.getPrice() > 250)
+                .toList();
+    }
+
+    public static List<Product> getBooksWithDiscount(List<Product> products) {
+        Objects.requireNonNull(products);
+        return products.stream()
+                .filter(product -> product.isBook() && product.hasDiscount())
+                .map(product -> {
+                    product.setPrice(product.getPrice() * 0.9);
+                    return product;
+                })
+                .toList();
+    }
+
+    public static Product findCheapestBook(List<Product> products) {
+        Objects.requireNonNull(products);
+        return products.stream()
+                .filter(Product::isBook)
+                .min(Comparator.comparingDouble(Product::getPrice))
+                .orElseThrow(() -> new RuntimeException("Product \"Book\" was not found!"));
+    }
+
+    public static List<Product> getLastThreeSupply(List<Product> products) {
+        Objects.requireNonNull(products);
+        return products.stream()
+                .collect(Collectors.toCollection(
+                        () -> new TreeSet<>(Comparator.comparing(Product::getSupplyDate).reversed())))
+                .stream()
+                .limit(3)
```

Comment:
Ok, it's good that you have found another solution, but there is a way to do that simpler. Could you please try to use `skip()` method of the Stream API? The idea is to skip all elements except the last three, as required in the task.